File: app/src/main/java/com/suhang/networkmvp/ui/fragment/HomeFragment.java
package com.suhang.networkmvp.ui.fragment;

import android.support.v7.widget.LinearLayoutManager;

import com.suhang.networkmvp.R;
import com.suhang.networkmvp.adapter.HomeRvAdapter;
import com.suhang.networkmvp.annotation.FragmentScope;
import com.suhang.networkmvp.dagger.module.BlankModule;
import com.suhang.networkmvp.databinding.FragmentHomeBinding;
import com.suhang.networkmvp.domain.DeleteHistoryBean;
import com.suhang.networkmvp.domain.HomeBean;
import com.suhang.networkmvp.mvp.event.ClickEvent;
import com.suhang.networkmvp.mvp.event.ItemClickEvent;
import com.suhang.networkmvp.mvp.model.HomeModel;
import com.suhang.networkmvp.mvp.result.ErrorResult;
import com.suhang.networkmvp.mvp.result.SuccessResult;
import com.suhang.networkmvp.utils.LogUtil;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

/**
 * Created by 苏杭 on 2017/1/24 15:31.
 */
@FragmentScope
public class HomeFragment extends BaseFragment<HomeModel,FragmentHomeBinding> {
    @Inject
    HomeRvAdapter mAdapter;
    public static final int TAG = 100;
    public static final int TAG_LOADMORE = 101;
    public static final int TAG_DELETE = 102;
    public static final int TAG_LOADMORE_NORMAL = 103;
    private List<Integer> mLs1 = new ArrayList<>();
    private List<String> mLs2 = new ArrayList<>();

    @Override
    protected int bindLayout() {
        return R.layout.fragment_home;
    }

    @Override
    protected void subscribeEvent() {
        getSm().subscribeEvent(ItemClickEvent.class).subscribe(itemClickEvent -> {

        });
        getSm().subscribeEvent(ClickEvent.class).subscribe(clickEvent -> {
            switch (clickEvent.getId()) {
                case R.id.button:
                    getModel().getLoadMore(mAdapter.getNextPage());
                    break;
                case R.id.button1:
                    LogUtil.i("啊啊啊"+mLs1+"   "+mLs2);
                    getModel().deleteHistory(mLs2, mLs1);
                    break;
            }
        });
        getSm().subscribeResult(ErrorResult.class).subscribe(errorResult -> {
            LogUtil.i("啊啊啊" + errorResult.getResult());
        });
        getSm().subscribeResult(SuccessResult.class).subscribe(successResult -> {
            if (successResult.getTag() == TAG_LOADMORE) {
                HomeBean result = successResult.getResult(HomeBean.class);
                mAdapter.setTotalCount(Integer.parseInt(result.getTotal()));
                mAdapter.loadMore(result.getList());
            } else if (successResult.getTag() == TAG) {
                HomeBean result = successResult.getResult(HomeBean.class);
                mAdapter.setTotalCount(Integer.parseInt(result.getTotal()));
                mAdapter.notifyDataSetChanged(result.getList());
            } else if (successResult.getTag() == TAG_DELETE) {
                DeleteHistoryBean result = successResult.getResult(DeleteHistoryBean.class);
                if (result.getFailedList() != null && "".equals(result.getFailedList())) {
                    getModel().getHomeData(mAdapter.getCurrentPage() * mAdapter.getPageSize(), (List<Integer>) result.getAppendMessage());
                }
            } else if (successResult.getTag() == TAG_LOADMORE_NORMAL) {
                HomeBean result = successResult.getResult(HomeBean.class);
                mAdapter.setTotalCount(Integer.parseInt(result.getTotal()));
                mAdapter.notifyDelete((List<Integer>) result.getAppendMessage(), result.getList());
                mLs1.clear();
                mLs2.clear();
            }

        });

    }

    @Override
    protected void initData() {
        mBinding.rvHome.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
        mBinding.rvHome.setAdapter(mAdapter);
        mModel.getHomeData();
    }

    @Override
    protected void injectDagger() {
        getBaseComponent().getBlankComponent(new BlankModule()).inject(this);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mAdapter.destory();
    }
}

File: app/src/main/java/com/suhang/networkmvp/ui/fragment/HomeFragment.kt
package com.suhang.networkmvp.ui.fragment

import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import com.suhang.layoutfinderannotation.GenSubComponent
import com.suhang.networkmvp.R
import com.suhang.networkmvp.adapter.HomeRvAdapter
import com.suhang.networkmvp.application.DaggerHelper
import com.suhang.networkmvp.constants.*
import com.suhang.networkmvp.dagger.module.ChildModule
import com.suhang.networkmvp.dagger.module.HomeModule
import com.suhang.networkmvp.domain.DeleteHistoryBean
import com.suhang.networkmvp.domain.HistoryBean
import com.suhang.networkmvp.mvp.model.HomeModel
import io.reactivex.functions.Consumer
import kotlinx.android.synthetic.main.fragment_home.*
import org.jetbrains.anko.warn
import javax.inject.Inject

/**
 * Created by 苏杭 on 2017/1/24 15:31.
 */
@GenSubComponent(tag = Constant.BASE_CHILD_DAGGER_TAG, modules = arrayOf(HomeModule::class))
class HomeFragment : BaseFragment<HomeModel>() {
    override fun initDagger() {
        DaggerHelper.getInstance().getHomeFragmentComponent(this, activity)
    }

    @Inject
    lateinit var mAdapter: HomeRvAdapter
    val luids: ArrayList<String> = ArrayList()
    val positions: ArrayList<Int> = ArrayList()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bind(R.layout.fragment_home)
    }

    override fun subscribeEvent() {
        manager.subscribeError().subscribe(Consumer {
            warn(it.result)
        })

        manager.subscribeSuccess().subscribe(Consumer {
            if (it.tag == TAG_LOADMORE) {
                (it.result as HistoryBean).run {
                    mAdapter.totalCount = total.toInt()
                    mAdapter.loadMore(list)
                }
            } else if (it.tag == TAG) {
                (it.result as HistoryBean).run {
                    mAdapter.totalCount = total.toInt()
                    mAdapter.notifyDataSetChanged(list)
                }
            } else if (it.tag == TAG_DELETE) {
                (it.result as DeleteHistoryBean).run {
                    if (failedList != null && "" == failedList) {
                        model.getHomeData(mAdapter.currentPage * mAdapter.pageSize, it.append as List<Int>)
                    }
                }
            } else if (it.tag == TAG_LOADMORE_NORMAL) {
                (it.result as HistoryBean).run {
                    mAdapter.totalCount = total.toInt()
                    mAdapter.notifyDelete(it.append as List<Int>, list)
                }
            }
        })

        manager.subscribeEvent().subscribe(Consumer {
            (it.view).run {
                setBackgroundColor(0xffff00ff.toInt())
            }
            (it.view.getAdapterTag() as HomeRvAdapter.Message).run {
                luids.add(luid)
                positions.add(position)
            }
        })

    }

    override fun initEvent() {
        button.setOnClickListener {
            model.getLoadMore(mAdapter.nextPage)
        }
        button1.setOnClickListener {
            model.deleteHistory(luids, positions)
        }
    }


    override fun initData() {
        rv_home.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        rv_home.adapter = mAdapter
        model.getHomeData()
    }

    override fun onDestroy() {
        super.onDestroy()
        mAdapter.destory()
    }

    companion object {
        val TAG = 101
        val TAG_LOADMORE = 102
        val TAG_DELETE = 103
        val TAG_LOADMORE_NORMAL = 104
    }
}
